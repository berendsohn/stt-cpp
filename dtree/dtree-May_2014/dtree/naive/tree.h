// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
