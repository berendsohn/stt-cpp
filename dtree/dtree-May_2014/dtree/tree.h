// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
