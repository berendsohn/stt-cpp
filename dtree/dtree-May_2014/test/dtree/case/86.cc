#define DTREE_CONFIG EndTree<WithEvert<WithAncAggr<dtree::DpMin<double>, dtree::DwLess_<double>, WithStaticValue<dtree::DpValue<double>, WithAncDescValue<FreeGroup, Begin > > > > >
#include "test/dtree/tree.cc"
