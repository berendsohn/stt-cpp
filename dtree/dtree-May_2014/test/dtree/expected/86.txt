<<<
  1: v650,000,000,000 vL4R-7F1 -> 6
  2: vb15,000,000,000 vL7R4F1 -> 13
  3: v065,000,000,000 vL-1R3F0
  4: v30b,000,000,000 vL6R-3F1 -> 7
  5: v5a9,000,000,000 vL-5R-5F1 -> 10
  6: vb18,000,000,000 vL-8R-5F0
  7: vd0d,000,000,000 vL2R5F1 -> 3
  8: vbf8,000,000,000 vL-4R2F0 -> 4
  9: vd1e,000,000,000 vL-7R-6F0 -> 7
  10: v2f4,000,000,000 vL-5R6F0 -> 11
  11: v01e,000,000,000 vL-5R-2F0 -> 1
  12: v26c,000,000,000 vL4R-4F0 -> 4
  13: vf30,000,000,000 vL0R0F0 -> 6
  14: v582,000,000,000 vL-6R6F1 -> 7
  15: v1dd,000,000,000 vL2R-5F0 -> 7
>>>
<<<
 7
    /12: v26c,30b,000,000 vL4R-4F0 aL4R-4 f0
  \4: v30b,000,000,000 vL6R-3F1 aL4R-4 f1

  @6: vb18,000,000,000 vL-8R-5F0 aL-8R-5 f0
    \1: v650,b18,000,000 vL4R-7F1 aL-5R-7 f0
        /5: v5a9,2f4,000,000 vL-5R-5F1 aL-5R-5 f1
      \10: v2f4,650,000,000 vL-5R6F0 aL-5R-5 f1
        \11: v01e,2f4,000,000 vL-5R-2F0 aL-5R-2 f0

 4
  \8: vbf8,000,000,000 vL-4R2F0 aL-4R2 f1

  @9: vd1e,000,000,000 vL-7R-6F0 aL-7R-6 f1
      /3: v065,d0d,000,000 vL-1R3F0 aL-1R3 f0
    \7: vd0d,d1e,000,000 vL2R5F1 aL-1R3 f1

 6
    /2: vb15,f30,000,000 vL7R4F1 aL7R4 f1
  \13: vf30,000,000,000 vL0R0F0 aL0R0 f1

 7
  \14: v582,000,000,000 vL-6R6F1 aL-6R6 f1

 7
  \15: v1dd,000,000,000 vL2R-5F0 aL2R-5 f0
>>>
